src/Features/LanguageServer/Protocol/Handler/LspMethodAttribute.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    /// <summary>
+    /// Defines a metadata attribute for <see cref="IRequestHandler{RequestType, ResponseType}"/>
+    /// to use to specify the kind of LSP request the handler accepts.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class), MetadataAttribute]
+    internal class LspMethodAttribute : Attribute, ILspMethodMetadata
+    {
+        /// <summary>
+        /// Name of the LSP method to handle.
+        /// </summary>
+        public string MethodName { get; }
+
+        /// <summary>
+        /// Whether or not handling this method results in changes to the current solution state.
+        /// Mutating requests will block all subsequent requests from starting until after they have
+        /// completed and mutations have been applied. See <see cref="RequestExecutionQueue"/>.
+        /// </summary>
+        public bool MutatesSolutionState { get; }
+
+        /// <summary>
+        /// If the <see cref="MethodName"/> is <see cref="Methods.WorkspaceExecuteCommandName"/>
+        /// then this is the name of the command to execute.
+        /// </summary>
+        public string? CommandName { get; }
+
+        public LspMethodAttribute(string methodName, bool mutatesSolutionState, string? commandName = null)
+        {
+            MethodName = methodName;
+            MutatesSolutionState = mutatesSolutionState;
+            CommandName = commandName;
+            if (Methods.WorkspaceExecuteCommandName.Equals(methodName, StringComparison.OrdinalIgnoreCase) && string.IsNullOrEmpty(CommandName))
```

Comment:
There's 2 reasons this is more difficult
1.  I can't easily share state between the normal handler and a workspace/executeCommand handler.  If I make a handler provider for execute command, it will require a different method (return different types of handlers) - so now creation of the code actions cache has to be shared between two different providers (or across 2 methods) which makes dealing with the lifetime more difficult
2.  All workspace/executeCommands would then have to become mutating.  Which seems like it would mostly be true?  But not sure I want to tie us down to that

Reply:
@davidwengier I left this using the same provider infrastructure (to resolve the issue about passing the same instance data to normal requests and commands), but I added a new attribute for commands to ensure we get a compile error if you do not specify a command name.  Let me know what you think

Reply:
Yeah, not being able to control mutation per command seems bad, thats a good point.

Reply:
I missed that the command we have needed the code action cache, that definitely complicates things.